org.osgi.test.assertj/src/main/java/org/osgi/test/assertj/bundle/AbstractBundleAssert.java

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) OSGi Alliance (2019). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.assertj.bundle;
+
+import static java.util.stream.Collectors.joining;
+import static org.osgi.framework.Bundle.ACTIVE;
+import static org.osgi.framework.Bundle.INSTALLED;
+import static org.osgi.framework.Bundle.RESOLVED;
+import static org.osgi.framework.Bundle.STARTING;
+import static org.osgi.framework.Bundle.STOPPING;
+import static org.osgi.framework.Bundle.UNINSTALLED;
+
+import java.net.URL;
+import java.time.Instant;
+import java.util.Date;
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+import org.assertj.core.api.AbstractAssert;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.Version;
+
+public abstract class AbstractBundleAssert<SELF extends AbstractBundleAssert<SELF, ACTUAL>, ACTUAL extends Bundle>
+	extends AbstractAssert<SELF, ACTUAL> {
+
+	public AbstractBundleAssert(ACTUAL actual, Class<SELF> selfType) {
+		super(actual, selfType);
+	}
+
+	public SELF isFragment() {
+		isNotNull();
+		String host = actual.getHeaders()
+			.get("Fragment-Host");
+		if (host == null) {
+			failWithMessage("%nExpecting%n  <%s>%nto be a fragment, but it was not", actual);
+		}
+		return myself;
+	}
+
+	public SELF isNotFragment() {
+		isNotNull();
+		String host = actual.getHeaders()
+			.get("Fragment-Host");
+		if (host != null) {
+			failWithMessage("%nExpecting%n  <%s>%nto not be a fragment, but it was, with host:%n  <%s>", actual, host);
+		}
+		return myself;
+	}
+
+	public SELF hasPermission(Object permission) {
+		isNotNull();
+		if (!actual.hasPermission(permission)) {
+			failWithMessage("%nExpecting%n  <%s>%nto have permission%n  <%s>%nbut it did not", actual, permission);
+		}
+		return myself;
+	}
+
+	public SELF doesNotHavePermission(Object permission) {
+		isNotNull();
+		if (actual.hasPermission(permission)) {
+			failWithMessage("%nExpecting%n  <%s>%nnot to have permission%n  <%s>%nbut it did", actual, permission);
+		}
+		return myself;
+	}
+
+	public SELF hasSymbolicName(String expected) {
+		isNotNull();
+		if (!Objects.equals(actual.getSymbolicName(), expected)) {
```

Comment:
There are pluses and minuses to either approach.

If you do it the way you've shown here, you'll get a `ComparisonFailure` rather than a bare `AssertionError` (or, if OpenTest4J is visible to the test classloader, then it will generate an `AssertionFailedError` using the three-arg constructor). This is useful in the GUI as this is how you get the visual diff (very useful for large strings). However, it means that if you use the assertion and (eg) write: 
`assertThat(bundle).as("my custom message").hasSymbolicName("my.bundle");`
... then the assertion message generated by this code won't include "my custom message". In fact, it won't honour any error message overrides.

The way I've done it here is the way the examples show on the AssertJ custom assertions page: https://joel-costigliola.github.io/assertj/assertj-core-custom-assertions.html. This way will honour the error message overrides, but it won't 

Personally, I've flip-flopped between both approaches. Ideally, there'd be an approach that allowed you to get the best of both worlds. The issue I recently raised against AssertJ is concerned precisely with this: joel-costigliola/assertj-core#1731

Reply:
Thanks for the info. There is no best way... yet.

Reply:
I figured out a solution which I used in PromiseAssert.

https://github.com/osgi/osgi-test/pull/73/files#diff-0cc790c08cc5cc8d0cec1e4e2418932fL243-R260

It converts the assert to an AbstractObjectAssert (with asInstanceOf) and then uses `extracting` to convert to assert to a new assert on the extracted datum. This propagates the assertion state from the original assert. (The conversion to AbstractObjectAssert is not needed if the assert is already a subtype.)

If `AbstractAssert.withAssertionState` was protected instead of package private, it would be preferred. But what I did is the best approximation with the public API now available.

Reply:
I've used your trick too.